README.md
# cache_proxy (BDP Cache Proxy Open Source Software (OSS) Edition) #
## Purpose ##
BDP Cache Proxy, aka **nutcracker'** (until someone wins the contest to name the service) is a
fast and lightweight proxy for [redis](http://redis.io/) protocol. It was built primarily to
reduce Riak KV mean response times, especially for immutable data, as well as reduce client
cost and implementation time while increasing flexibility in implementations by supporting with
little effort a horizontally scalable distributed cache and persistent data store architecture.

## Lineage ##
BDP Cache Proxy is based on twemproxy, with the following high-level differences:
1. Cache strategy implementations are moved from the client into the Cache Proxy service.
2. Upstream persistent data stores, aka 'Backend' servers were necessarily added.
2.a. Riak KV is the one and only persistent data store supported.
2.b. Redis may also be configured as a Backend server, useful only for development/testing.

As mentioned earlier, nutcracker' is subject to name change. All processes dependent upon the
executable's name are thus subject to change. For typical BDP use, all such changes, ie
BDP Service Manager (SM) integration, will be encapsulated in BDP.

See README.twemproxy.md for fuller twemproxy coverage, but consider any topic not covered here to
not be supported use of the BDP Cache Proxy.

## Dependencies ##
BDP Cache Proxy depends upon libprotobuf-c to communicate with Riak KV.  If you already have
libprotobuf-c present, skip to Build.

### Build libprotobuf-c ###
Refer to https://github.com/protobuf-c/protobuf-c

```bash
git clone https://github.com/protobuf-c/protobuf-c
cd protobuf-c
./autogen.sh && ./configure && make
make install
```

## Build ##
To build BDP Cache Proxy from source with _debug logs enabled_ and _assertions enabled_:

    $ git clone git@github.com:basho/cache_proxy.git
    $ cd cache_proxy
    $ autoreconf -fvi
    $ ./configure --enable-debug=full --with-protobuf-c=/usr/local
    $ make
    $ src/nutcracker -h

A quick checklist:

+ Use newer version of gcc (older version of gcc has problems)
+ Use CFLAGS="-O1" ./configure && make
+ Use CFLAGS="-O3 -fno-strict-aliasing" ./configure && make
+ `autoreconf -fvi && ./configure` needs `automake` and `libtool` to be installed

## Features ##

+ Fast.
+ Lightweight.
+ Maintains persistent server connections.
+ Keeps connection count on the caching and persistent data store servers low.
+ Enables pipelining of requests and responses.
+ Supports proxying to multiple servers.
+ Supports multiple server pools simultaneously.
+ Shard data automatically across multiple servers.
+ Easy configuration of server pools through a YAML file.
+ Supports multiple hashing modes including consistent hashing and distribution.
+ Can be configured to disable nodes on failures.
+ Observability via stats exposed on the stats monitoring port.
+ Works with Centos, Linux, *BSD, OS X.

## Help ##

    Usage: nutcracker [-?hVdDt] [-v verbosity level] [-o output file]
                      [-c conf file] [-s stats port] [-a stats addr]
                      [-i stats interval] [-p pid file] [-m mbuf size]

    Options:
      -h, --help             : this help
      -V, --version          : show version and exit
      -t, --test-conf        : test configuration for syntax errors and exit
      -d, --daemonize        : run as a daemon
      -D, --describe-stats   : print stats description and exit
      -v, --verbose=N        : set logging level (default: 5, min: 0, max: 11)
      -o, --output=S         : set logging file (default: stderr)
      -c, --conf-file=S      : set configuration file (default: conf/nutcracker.yml)
      -s, --stats-port=N     : set stats monitoring port (default: 22222)
      -a, --stats-addr=S     : set stats monitoring ip (default: 0.0.0.0)
      -i, --stats-interval=N : set stats aggregation interval in msec (default: 30000 msec)
      -p, --pid-file=S       : set pid file (default: off)
      -m, --mbuf-size=N      : set size of mbuf chunk in bytes (default: 16384 bytes)

## Configuration ##
BDP Cache Proxy can be configured through a YAML file specified by the -c or --conf-file
command-line argument on process start. The configuration file is used to specify the server
pools and the servers within each pool that the BDP Cache Proxy manages. The configuration files
supports the following keys:

+ **listen**: The listening address and port (name:port or ip:port) or an absolute path to sock file (e.g. /var/run/nutcracker.sock) for this server pool.
+ **hash**: The name of the hash function. Possible values are:
 + one_at_a_time
 + md5
 + crc16
 + crc32 (crc32 implementation compatible with [libmemcached](http://libmemcached.org/))
 + crc32a (correct crc32 implementation as per the spec)
 + fnv1_64
 + fnv1a_64
 + fnv1_32
 + fnv1a_32
 + hsieh
 + murmur
 + jenkins
+ **hash_tag**: A two character string that specifies the part of the key used for hashing. Eg "{}" or "$$". [Hash tag](notes/recommendation.md#hash-tags) enable mapping different keys to the same server as long as the part of the key within the tag is the same.
+ **distribution**: The key distribution mode. Possible values are:
 + ketama
 + modula
 + random
+ **timeout**: The timeout value in msec that we wait for to establish a connection to the server or receive a response from a server. By default, we wait indefinitely.
+ **backlog**: The TCP backlog argument. Defaults to 512.
+ **preconnect**: A boolean value that controls if the Cache Proxy should preconnect to all the
servers in this pool on process start. Defaults to false.
+ **redis**: A boolean value that controls if a server pool speaks redis or memcached protocol.
Defaults to false.
+ **redis_auth**: Authenticate to the Redis server on connect.
+ **server_connections**: The maximum number of connections that can be opened to each server. By default, we open at most 1 server connection.
+ **auto_eject_hosts**: A boolean value that controls if server should be ejected temporarily when it fails consecutively server_failure_limit times. See [liveness recommendations](notes/recommendation.md#liveness) for information. Defaults to false.
+ **server_retry_timeout**: The timeout value in msec to wait for before retrying on a temporarily ejected server, when auto_eject_host is set to true. Defaults to 30000 msec.
+ **server_failure_limit**: The number of consecutive failures on a server that would lead to it being temporarily ejected when auto_eject_host is set to true. Defaults to 2.
+ **server_ttl**: Cache time-to-live (TTL), specified in unit format, ie 15s for 15 seconds.
+ **servers**: A list of server address, port and weight (name:port:weight or ip:port:weight) for this server pool.
+ **backend_type**: riak (supported) or redis (useful only for development/testing)
+ **backend_max_resend**: positive integer, the maximum number of times to attempt resending a
message to the backend server.
+ **backends**: A list of server address, port and weight (name:port:weight or ip:port:weight) for this server pool.

For example, see the configuration file in [conf/backend_test.yml](conf/backend_test.yml).

Finally, to make writing a syntactically correct configuration file easier, BDP Cache Proxy
provides a command-line argument -t or --test-conf that can be used to test the YAML
configuration file for any syntax error.

## Observability

Observability in BDP Cache Proxy is through logs and stats.

BDP Cache Proxy exposes stats at the granularity of server pool and servers per pool through the stats monitoring port. The stats are essentially JSON formatted key-value pairs, with the keys corresponding to counter names. By default stats are exposed on port 22222 and aggregated every 30 seconds. Both these values can be configured on program start using the -c or --conf-file and -i or --stats-interval command-line arguments respectively. You can print the description of all stats exported by  using the -D or --describe-stats command-line argument.

    $ nutcracker --describe-stats

    pool stats:
      client_eof          "# eof on client connections"
      client_err          "# errors on client connections"
      client_connections  "# active client connections"
      server_ejects       "# times backend server was ejected"
      forward_error       "# times we encountered a forwarding error"
      fragments           "# fragments created from a multi-vector request"

    server stats:
      server_eof          "# eof on server connections"
      server_err          "# errors on server connections"
      server_timedout     "# timeouts on server connections"
      server_connections  "# active server connections"
      requests            "# requests"
      request_bytes       "total request bytes"
      responses           "# responses"
      response_bytes      "total response bytes"
      in_queue            "# requests in incoming queue"
      in_queue_bytes      "current request bytes in incoming queue"
      out_queue           "# requests in outgoing queue"
      out_queue_bytes     "current request bytes in outgoing queue"

Logging in BDP Cache Proxy is only available when built with logging enabled. By default logs are written to stderr. BDP Cache Proxy can also be configured to write logs to a specific file through the -o or --output command-line argument. On a running BDP Cache Proxy, we can turn log levels up and down by sending it SIGTTIN and SIGTTOU signals respectively and reopen log files by sending it SIGHUP signal.

## Pipelining

BDP Cache Proxy enables proxying multiple client connections onto one or few server connections. This architectural setup makes it ideal for pipelining requests and responses and hence saving on the round trip time.

For example, if BDP Cache Proxy is proxying three client connections onto a single server and we get requests - 'get key\r\n', 'set key 0 0 3\r\nval\r\n' and 'delete key\r\n' on these three connections respectively, BDP Cache Proxy would try to batch these requests and send them as a single message onto the server connection as 'get key\r\nset key 0 0 3\r\nval\r\ndelete key\r\n'.

Pipelining is the reason why BDP Cache Proxy ends up doing better in terms of throughput even though it introduces an extra hop between the client and server.

## Deployment

If you are deploying BDP Cache Proxy in production, you might consider reading through the [recommendation document](notes/recommendation.md) to understand the parameters you could tune in BDP Cache Proxy to run it efficiently in the production environment.  However, by installing BDP Cache Proxy as part of the Basho Data Platform, these recommendations have already been heeded, yielding the best performance and reducing the total cost of ownership.

## Packages

### Centos
#TODO: create Centos (RH, rpm) package

### Ubuntu
#TODO: create Ubuntu (Debian, deb) package

## Issues and Support

Have a bug or a question? Please create an issue here on GitHub!

https://github.com/basho/cache_proxy/issues

## Contribution ##
Thank you to all of our [contributors](https://github.com/basho/cache_proxy/graphs/contributors)!

## License

© 2015 Basho Technologies

Licensed under the Apache License, Version 2.0: http://www.apache.org/licenses/LICENSE-2.0


ChangeLog
  2015-08-21  paegun <l2paegun@hotmail.com>
    * cache_proxy: version 1.0.0, fork from twemproxy 0.4.0
      added backend server, supporting only Riak
      added read-through caching strategy for GET command
      this change is a rollup of the research and development of
	  several contributors including:
        aberghage <aberghage@gmail.com>
        erikleitch <eleitch@basho.com>
        paegun <l2paegun@hotmail.com>
        Nikolay-Kha <Nikolay.Khabarov@dataart.com>

2014-18-10  idning <idning@gmail.com>
    * twemproxy: version 0.4.0 release
      mget improve (idning)
      many new commands supported: LEX, PFADD, PFMERGE, SORT, PING, QUIT, SCAN... (mattrobenolt, areina, idning)
      handle max open file limit(allenlz)
      add notice-log and use ms time in log(idning)
      fix bug in string_compare (andyqzb)
      fix deadlock in sighandler (idning)

2013-20-12  Manju Rajashekhar  <manj@cs.stanford.edu>
    * twemproxy: version 0.3.0 release
      SRANDMEMBER support for the optional count argument (mkhq)
      Handle case where server responds while the request is still being sent (jdi-tagged)
      event ports (solaris/smartos) support
      add timestamp when the server was ejected
      support for set ex/px/nx/xx for redis 2.6.12 and up (ypocat)
      kqueue (bsd) support (ferenyx)
      fix parsing redis response to accept integer reply (charsyam)

2013-23-04  Manju Rajashekhar  <manj@cs.stanford.edu>
    * twemproxy: version 0.2.4 release
      redis keys must be less than mbuf_data_size() in length (fifsky)
      Adds support for DUMP/RESTORE commands in Redis (remotezygote)
      Use of the weight value in the modula distribution (mezzatto)
      Add support to unix socket connections to servers (mezzatto)
      only check for duplicate server name and not 'host:port:weight' when 'name' is configured
      crc16 hash support added (mezzatto)

2013-31-01  Manju Rajashekhar  <manj@twitter.com>
    * twemproxy: version 0.2.3 release
      RPOPLPUSH, SDIFF, SDIFFSTORE, SINTER, SINTERSTORE, SMOVE, SUNION, SUNIONSTORE, ZINTERSTORE, and ZUNIONSTORE support (dcartoon)
      EVAL and EVALSHA support (ferenyx)
      exit 1 if configuration file is invalid (cofyc)
      return non-zero exit status when nutcracker cannot start for some reason
      use server names in stats (charsyam)
      Fix failure to resolve long FQDN name resolve (conmame)
      add support for hash tags

2012-18-10  Manju Rajashekhar  <manj@twitter.com>

    * twemproxy: version 0.2.2 release
      fix the off-by-one error when calculating redis key length

2012-12-10  Manju Rajashekhar  <manj@twitter.com>

    * twemproxy: version 0.2.1 release
      don't use buf in conf_add_server
      allow an optional instance name for consistent hashing (charsyam)
      add --stats-addr=S option
      add stats-bind-any -a option (charsyam)

2012-12-03  Manju Rajashekhar  <manj@twitter.com>

    * twemproxy: version 0.2.0 release
      add -D or --describe-stats command-line argument to print stats description
      redis support in twemproxy
      setup pre/post splitcopy and pre/post coalesce handlers in msg struct
      memcache pre_splitcopy, post_splitcopy, pre_coalesce and post_coalesce handlers
      every fragment of a msg vector keeps track of the first/last fragment, number of fragments and fragment owner
      set up msg parser handler for memcache connections
      refactor parsing code and create header file nc_proto.h
      stats_listen should use st->addr as the listening address string
      delete stats tracking memcache requests and responses; stats module no longer tracks protocol related stats

2012-10-27  Manju Rajashekhar  <manj@twitter.com>

    * twemproxy: version 0.1.20 release
      on msg_repair, msg->pos should point to nbuf->pos and not nbuf->last
      refactor memcache parsing code into proto directory
      add redis option to configuration file
      fix macro definition strXcmp error for big endian
      fix log_hexdump and loga_hexdump

2012-07-31  Manju Rajashekhar  <manj@twitter.com>

    * twemproxy: version 0.1.19 release
      close server connection on a stray response (yashh, bmatheny)

2012-06-19  Manju Rajashekhar  <manj@twitter.com>

    * twemproxy: version 0.1.18 release
      command line option to set mbuf chunk size

2012-05-09  Manju Rajashekhar  <manj@twitter.com>

    * twemproxy: version 0.1.17 release
      use _exit(0) instead of exit(0) when daemonizing
      use loga instead of log_stderr in nc_stacktrace

2012-02-09  Manju Rajashekhar  <manj@twitter.com>

    * twemproxy: version 0.1.16 release
      twemproxy (aka nutcracker) is a fast and lightweight proxy for memcached protocol.



src/nc_stats.h
/*
 * twemproxy - A fast and lightweight proxy for memcached protocol.
 * Copyright (C) 2011 Twitter, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#ifndef _NC_STATS_H_
#define _NC_STATS_H_

#include <nc_core.h>

#define STATS_POOL_CODEC(ACTION)                                                                                    \
    /* client behavior */                                                                                           \
    ACTION( client_eof,             STATS_COUNTER,      "# eof on client connections")                              \
    ACTION( client_err,             STATS_COUNTER,      "# errors on client connections")                           \
    ACTION( client_connections,     STATS_GAUGE,        "# active client connections")                              \
    /* pool behavior */                                                                                             \
    ACTION( server_ejects,          STATS_COUNTER,      "# times backend server was ejected")                       \
    /* forwarder behavior */                                                                                        \
    ACTION( forward_error,          STATS_COUNTER,      "# times we encountered a forwarding error")                \
    ACTION( fragments,              STATS_COUNTER,      "# fragments created from a multi-vector request")          \

#define STATS_SERVER_CODEC(ACTION)                                                                                  \
    /* server behavior */                                                                                           \
    ACTION( server_eof,             STATS_COUNTER,      "# eof on server connections")                              \
    ACTION( server_err,             STATS_COUNTER,      "# errors on server connections")                           \
    ACTION( server_timedout,        STATS_COUNTER,      "# timeouts on server connections")                         \
    ACTION( server_connections,     STATS_GAUGE,        "# active server connections")                              \
    ACTION( server_ejected_at,      STATS_TIMESTAMP,    "timestamp when server was ejected in usec since epoch")    \
    /* data behavior */                                                                                             \
    ACTION( requests,               STATS_COUNTER,      "# requests")                                               \
    ACTION( request_bytes,          STATS_COUNTER,      "total request bytes")                                      \
    ACTION( responses,              STATS_COUNTER,      "# responses")                                              \
    ACTION( response_bytes,         STATS_COUNTER,      "total response bytes")                                     \
    ACTION( in_queue,               STATS_GAUGE,        "# requests in incoming queue")                             \
    ACTION( in_queue_bytes,         STATS_GAUGE,        "current request bytes in incoming queue")                  \
    ACTION( out_queue,              STATS_GAUGE,        "# requests in outgoing queue")                             \
    ACTION( out_queue_bytes,        STATS_GAUGE,        "current request bytes in outgoing queue")                  \

#define STATS_ADDR      "0.0.0.0"
#define STATS_PORT      22222
#define STATS_INTERVAL  (30 * 1000) /* in msec */

typedef enum stats_type {
    STATS_INVALID,
    STATS_COUNTER,    /* monotonic accumulator */
    STATS_GAUGE,      /* non-monotonic accumulator */
    STATS_TIMESTAMP,  /* monotonic timestamp (in nsec) */
    STATS_SENTINEL
} stats_type_t;

struct stats_metric {
    stats_type_t  type;         /* type */
    struct string name;         /* name (ref) */
    union {
        int64_t   counter;      /* accumulating counter */
        int64_t   timestamp;    /* monotonic timestamp */
    } value;
};

struct stats_server {
    struct string name;   /* server name (ref) */
    struct array  metric; /* stats_metric[] for server codec */
};

struct stats_pool {
    struct string name;   /* pool name (ref) */
    struct array  metric; /* stats_metric[] for pool codec */
    struct array  server; /* stats_server[] */
    struct array  server_be; /* stats_server_be[] */
};

struct stats_buffer {
    size_t   len;   /* buffer length */
    uint8_t  *data; /* buffer data */
    size_t   size;  /* buffer alloc size */
};

struct stats {
    uint16_t            port;            /* stats monitoring port */
    int                 interval;        /* stats aggregation interval */
    struct string       addr;            /* stats monitoring address */

    int64_t             start_ts;        /* start timestamp of nutcracker */
    struct stats_buffer buf;             /* output buffer */

    struct array        current;         /* stats_pool[] (a) */
    struct array        shadow;          /* stats_pool[] (b) */
    struct array        sum;             /* stats_pool[] (c = a + b) */

    pthread_t           tid;             /* stats aggregator thread */
    int                 sd;              /* stats descriptor */

    struct string       service_str;     /* service string */
    struct string       service;         /* service */
    struct string       source_str;      /* source string */
    struct string       source;          /* source */
    struct string       version_str;     /* version string */
    struct string       version;         /* version */
    struct string       uptime_str;      /* uptime string */
    struct string       timestamp_str;   /* timestamp string */
    struct string       ntotal_conn_str; /* total connections string */
    struct string       ncurr_conn_str;  /* curr connections string */

    volatile int        aggregate;       /* shadow (b) aggregate? */
    volatile int        updated;         /* current (a) updated? */
};

#define DEFINE_ACTION(_name, _type, _desc) STATS_POOL_##_name,
typedef enum stats_pool_field {
    STATS_POOL_CODEC(DEFINE_ACTION)
    STATS_POOL_NFIELD
} stats_pool_field_t;
#undef DEFINE_ACTION

#define DEFINE_ACTION(_name, _type, _desc) STATS_SERVER_##_name,
typedef enum stats_server_field {
    STATS_SERVER_CODEC(DEFINE_ACTION)
    STATS_SERVER_NFIELD
} stats_server_field_t;
#undef DEFINE_ACTION

#if defined NC_STATS && NC_STATS == 1

#define stats_pool_incr(_ctx, _pool, _name) do {                        \
    _stats_pool_incr(_ctx, _pool, STATS_POOL_##_name);                  \
} while (0)

#define stats_pool_decr(_ctx, _pool, _name) do {                        \
    _stats_pool_decr(_ctx, _pool, STATS_POOL_##_name);                  \
} while (0)

#define stats_pool_incr_by(_ctx, _pool, _name, _val) do {               \
    _stats_pool_incr_by(_ctx, _pool, STATS_POOL_##_name, _val);         \
} while (0)

#define stats_pool_decr_by(_ctx, _pool, _name, _val) do {               \
    _stats_pool_decr_by(_ctx, _pool, STATS_POOL_##_name, _val);         \
} while (0)

#define stats_pool_set_ts(_ctx, _pool, _name, _val) do {                \
    _stats_pool_set_ts(_ctx, _pool, STATS_POOL_##_name, _val);          \
} while (0)

#define stats_server_incr(_ctx, _server, _name) do {                    \
    _stats_server_incr(_ctx, _server, STATS_SERVER_##_name);            \
} while (0)

#define stats_server_decr(_ctx, _server, _name) do {                    \
    _stats_server_decr(_ctx, _server, STATS_SERVER_##_name);            \
} while (0)

#define stats_server_incr_by(_ctx, _server, _name, _val) do {           \
    _stats_server_incr_by(_ctx, _server, STATS_SERVER_##_name, _val);   \
} while (0)

#define stats_server_decr_by(_ctx, _server, _name, _val) do {           \
    _stats_server_decr_by(_ctx, _server, STATS_SERVER_##_name, _val);   \
} while (0)

#define stats_server_set_ts(_ctx, _server, _name, _val) do {            \
     _stats_server_set_ts(_ctx, _server, STATS_SERVER_##_name, _val);   \
} while (0)

#else

#define stats_pool_incr(_ctx, _pool, _name)

#define stats_pool_decr(_ctx, _pool, _name)

#define stats_pool_incr_by(_ctx, _pool, _name, _val)

#define stats_pool_decr_by(_ctx, _pool, _name, _val)

#define stats_server_incr(_ctx, _server, _name)

#define stats_server_decr(_ctx, _server, _name)

#define stats_server_incr_by(_ctx, _server, _name, _val)

#define stats_server_decr_by(_ctx, _server, _name, _val)

#endif

#define stats_enabled   NC_STATS

void stats_describe(void);

void _stats_pool_incr(struct context *ctx, struct server_pool *pool, stats_pool_field_t fidx);
void _stats_pool_decr(struct context *ctx, struct server_pool *pool, stats_pool_field_t fidx);
void _stats_pool_incr_by(struct context *ctx, struct server_pool *pool, stats_pool_field_t fidx, int64_t val);
void _stats_pool_decr_by(struct context *ctx, struct server_pool *pool, stats_pool_field_t fidx, int64_t val);
void _stats_pool_set_ts(struct context *ctx, struct server_pool *pool, stats_pool_field_t fidx, int64_t val);

void _stats_server_incr(struct context *ctx, struct server *server, stats_server_field_t fidx);
void _stats_server_decr(struct context *ctx, struct server *server, stats_server_field_t fidx);
void _stats_server_incr_by(struct context *ctx, struct server *server, stats_server_field_t fidx, int64_t val);
void _stats_server_decr_by(struct context *ctx, struct server *server, stats_server_field_t fidx, int64_t val);
void _stats_server_set_ts(struct context *ctx, struct server *server, stats_server_field_t fidx, int64_t val);

struct stats *stats_create(uint16_t stats_port, char *stats_ip, int stats_interval, char *source, struct array *server_pool);
void stats_destroy(struct stats *stats);
void stats_swap(struct stats *stats);

#endif
